app/ModelFunctions/AlbumFunctions.php

```diff
@@ -340,4 +348,55 @@ public function get_sub_albums($parentAlbum, $return = array())
 
 		return $return;
 	}
+
+	/**
+	 * Returns an array of top-level albums and shared albums accessible by
+	 * the current user.
+	 *
+	 * @return array or null
+	 */
+	public function getToplevelAlbums()
+	{
+		$return = array(
+			'albums' => null,
+			'shared_albums' => null,
+		);
+
+		if ($this->sessionFunctions->is_logged_in()) {
+			$id = $this->sessionFunctions->id();
+			$user = User::find($id);
+
+			if ($id == 0) {
+				$return['albums'] = Album::where('owner_id', '=', 0)
```

Comment:
aren't we missing a `with(['children'])` here?

Reply:
Pardon my ignorance, but I never actually understood what those `with` clauses do so I honestly can't answer your question :-(. All I can say is that I just moved this code from the `AlbumsController`, so that's the way it was there and it seemed to work fine :smile:.

Reply:
Have a look at that: https://laravel.com/docs/5.8/eloquent-relationships#eager-loading :)

Reply:
Ah, now I understand why I couldn't understand them -- they have no semantic meaning, they are optimization hints :smile:. Right, so I guess `children` is likely to come in handy given that the callers will travel down the subalbum tree. Should we indicate `owner` as well, like we do for shared albums? I'm guessing it may come in handy when checking permissions using `readAccessFunctions->album()`?

Reply:
OK, I tried adding `with(['children'])` and also `with(['owner', 'children'])` but I failed to notice any significant difference in the total number of SQL queries performed. Is it still worth adding then?